modules/fedimint-mint-client/src/backup/tracker.rs

```diff
@@ -0,0 +1,716 @@
+
+    /// `LastECashNoteIndex` but tracked in flight. Basically max index of any
+    /// note that got a partial sig from the federation (initialled from the
+    /// backup value). TODO: One could imagine a case where the note was
+    /// issued but not get any partial sigs yet. Very unlikely in real life
+    /// scenario, but worth considering.
+    last_mined_nonce_idx: Tiered<NoteIndex>,
+
+    /// Threshold
+    threshold: u64,
+
+    /// Public key shares for each peer
+    ///
+    /// Used to validate contributed consensus items
+    pub_key_shares: BTreeMap<PeerId, Tiered<PublicKeyShare>>,
+
+    /// Aggregate public key for each amount tier
+    tbs_pks: Tiered<AggregatePublicKey>,
+
+    /// The number of nonces we look-ahead when looking for mints (per each
+    /// amount).
+    gap_limit: u64,
+}
+
+impl MintRestoreInProgressState {
+    fn transitions(
+        &self,
+        operation_id: OperationId,
+        context: &MintClientContext,
+        global_context: &DynGlobalClientContext,
+    ) -> Vec<StateTransition<MintRestoreStateMachine>> {
+        // TODO: pass refund contexts into state transitions
+        let make_progress_gc = global_context.clone();
+        let timeout_cancel_gc = global_context.clone();
+        let mut state = self.clone();
+        vec![StateTransition::new(
+            self.make_progress(
+                make_progress_gc.api(),
+                &global_context.decoders(),
+                context.epoch_pk,
+                context.secret.clone(),
+            ),
+            move |dbtx, new_state, old_state_machine: MintRestoreStateMachine| {
+                Box::pin(async move {
+                    if new_state.is_done() {
+                        MintRestoreStateMachine {
+                            operation_id: old_state_machine.operation_id,
+                            state: MintRestoreStates::Success,
+                        }
+                    } else {
+                        MintRestoreStateMachine {
+                            operation_id: old_state_machine.operation_id,
+                            state: MintRestoreStates::InProgress(new_state),
+                        }
+                    }
+                })
+            },
+        )]
+    }
+
+    async fn make_progress<'a>(
+        mut self,
+        api: &'a (dyn IFederationApi + 'static),
+        decoders: &'a ModuleDecoderRegistry,
+        epoch_pk: threshold_crypto::PublicKey,
+        secret: DerivableSecret,
+    ) -> Self {
+        assert_eq!(secret.level(), 1);
+        let epoch_range =
+            self.next_epoch..cmp::min(self.next_epoch.wrapping_add(100), self.end_epoch);
+        let mut epoch_stream = Self::fetch_epochs_stream(api, epoch_pk, decoders, epoch_range);
+        while let Some((epoch, epoch_history)) = epoch_stream.next().await {
+            // TODO: should `GlobalClientContext` expose a `TaskHandle` to allow us to react
+            // to shutdown?
```

Comment:
No, I think the state machine executor would just interrupt the trigger future (needs to be tested).